### PR TITLE
Name `BulkItemResponse` ctors

### PR DESCRIPTION
**Pull request by nik9000**
`BulkItemResponse` can contain either a success or failure. This
replaces the two constructors used to build either case with named
static methods. So instead of
```
return new BulkItemResponse(0, OpType.CREATE, createResponse);
return new BulkItemResponse(0, OpType.CREATE, failure);
```
you now use
```
return BulkItemResponse.success(0, OpType.CREATE, createResponse);
return BulkItemResponse.failure(0, OpType.CREATE, failure);
```

This makes it marginally easier to read code building these things - you
don't have to know the type of the parameter to know if its a failure
or success.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java [hunk 1]
```diff
@@ -341,66 +341,41 @@ public String toString() {
...
+        response = readResponse(in);
+        failure = in.readBoolean() ? new Failure(in) : null;
     }
 
-    public BulkItemResponse(int id, OpType opType, DocWriteResponse response) {
+    private BulkItemResponse(int id, OpType opType, DocWriteResponse response, Failure failure) {
         this.id = id;
         this.response = response;
         this.opType = opType;
+        this.failure = failure;
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java:370]: I think it makes sense to add an assertion here that either `response` or `failure` needs to be set, but not both.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java [hunk 2]
```diff
@@ -341,66 +341,41 @@ public String toString() {
...
-    private OpType opType;
+    private final OpType opType;
 
-    private DocWriteResponse response;
+    private final DocWriteResponse response;
 
-    private Failure failure;
-
-    BulkItemResponse() {}
+    private final Failure failure;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java:350]: This all started with me wanting to make these final to make them easier to reason about.

**Review by arteam — APPROVED** [review]
LGTM! Left one small comment

File: server/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java:370]: :+1:

**Comment by nik9000** [comment]
Thanks for reviewing @arteam.

**Comment by nik9000** [comment]
Manual backport incoming.

**Comment by nik9000** [comment]
High touch backport! Just about every line conflicted because master doesn't have types. Fun.

**Pull request merged**